src/docker/restorableappengine.cc

```diff
@@ -47,30 +73,67 @@ bool RestorableAppEngine::pullImages(const App& app) {
   return true;
 }
 
-bool RestorableAppEngine::installApp(const App& app) {
+bool RestorableAppEngine::installAppWithDocker(const App& app) {
   Docker::Uri uri{Docker::Uri::parseUri(app.uri)};
   const std::string archive_file_name{uri.digest.shortHash() + '.' + app.name + ArchiveExt};
-  const auto restore_app_root = RestorableAppEngine::appRoot(app);
-  extractAppArchive(app, (restore_app_root / archive_file_name).string());
+  extractAppArchive(app, (apps_root_ / app.name / archive_file_name).string());
+  return true;
+}
 
-  return ComposeAppEngine::installApp(app);
+bool cmd_str(const std::string& cmd) {
+  LOG_DEBUG << "Running: " << cmd;
+  int exit_code = boost::process::system(cmd);
+  return exit_code == 0;
 }
 
-bool RestorableAppEngine::start(const App& app) {
-  Docker::Uri uri{Docker::Uri::parseUri(app.uri)};
-  const std::string archive_file_name{uri.digest.shortHash() + '.' + app.name + ArchiveExt};
-  const auto restore_app_root = root_ / "apps" / app.name;
-  extractAppArchive(app, (restore_app_root / archive_file_name).string());
+bool skopeo_copy(const std::string& auth, const std::string& src_uri, const std::string& dst_blob_path,
+                 const std::string& dst_path) {
+  std::string cmd{"skopeo copy -f v2s2 --src-creds=" + auth + " --dest-shared-blob-dir " + dst_blob_path +
+                  " docker://" + src_uri + " oci:" + dst_path};
+  LOG_ERROR << "???? " << cmd;
+  return cmd_str(cmd);
+}
 
-  return ComposeAppEngine::start(app);
+bool RestorableAppEngine::pullImagesWithSkopeo(const App& app) {
+  std::string fio_hub_auth;
+  use_restore_root_ = true;
+  ComposeInfo compose{(appRoot(app) / ComposeAppEngine::ComposeFile).string()};
+  for (const auto& service : compose.getServices()) {
+    const auto image = compose.getImage(service);
+    LOG_ERROR << ">>>" << image;
+    const Uri uri{Uri::parseUri(image)};
+    std::string auth{"anonymous:"};
+
+    if (uri.registryHostname == "hub.foundries.io") {
```

Comment:
we do have one customer that's looking to using a private registry (AWS). Could skopeo-login help with this in a generic way?

Reply:
The regular docker cred helpers should work, it works for "hub.foundries.io" I just wanted to optimize and bypass the aktualizr based helper which is called on each image pull. What is need it in addition to this code is to add `export REGISTRY_AUTH_FILE=/usr/lib/docker/config.json` to the aklite service recipe in meta-lmp.

Reply:
so, customers can define cred helpers to their registries and refer to them in `/usr/lib/docker/config.json`.

Reply:
as result, it will work for both type of app engines, restorable and the regular one.